addons/auth_totp_mail_enforce/models/res_users.py

```diff
@@ -0,0 +1,138 @@
+        self._totp_rate_limit_purge('code_check')
+        return True
+
+    def _get_totp_mail_key(self):
+        self.ensure_one()
+        secret = self.env['ir.config_parameter'].sudo().get_param('database.secret')
+        assert secret, "Mail TOTP protection requires a configured database secret"
+        return f'{self.id}/{self.login}/{secret}'.encode()
+
+    def _get_totp_mail_code(self):
+        self.ensure_one()
+
+        key = self._get_totp_mail_key()
+
+        now = datetime.now()
+        counter = int(datetime.timestamp(now) / 3600)
+
+        code = hotp(key, counter)
+        expiration = timedelta(seconds=3600)
+        lang = babel_locale_parse(self.env.context.get('lang') or self.lang)
+        expiration = babel.dates.format_timedelta(expiration, lang)
+
+        return code, expiration
+
+    def _send_totp_mail_code(self):
+        self.ensure_one()
+        self._totp_rate_limit('send_email')
+
+        template = self.env.ref('auth_totp_mail_enforce.mail_template_totp_mail_code').sudo()
+        context = {}
+        if request:
+            geoip = request.session.geoip
+            context.update({
+                'location': f"{geoip['city']}, {geoip['country_name']}" if geoip else None,
+                'device': request.httprequest.user_agent.platform,
+                'browser': request.httprequest.user_agent.browser,
+                'ip': request.httprequest.environ['REMOTE_ADDR'],
+            })
+        email_values = {
+            'email_to': self.email,
+            'email_cc': False,
+            'auto_delete': True,
+            'recipient_ids': [],
+            'partner_ids': [],
+            'scheduled_date': False,
+        }
+        with self.env.cr.savepoint():
+            template.with_context(**context).send_mail(
+                self.id, force_send=True, raise_exception=True, email_values=email_values, notif_layout='mail.mail_notification_light'
+            )
+
+    def _totp_rate_limit(self, limit_type):
+        self.ensure_one()
+        assert request, "A request is required to be able to rate limit TOTP related actions"
+        limit, interval = TOTP_RATE_LIMITS.get(limit_type)
+        RateLimitLog = self.env['auth.totp.rate.limit.log'].sudo()
+        ip = request.httprequest.environ['REMOTE_ADDR']
+        domain = [
+            ('user_id', '=', self.id),
+            ('create_date', '>=', datetime.now() - timedelta(seconds=interval)),
+            ('limit_type', '=', limit_type),
+            ('ip', '=', ip),
+        ]
+        count = RateLimitLog.search_count(domain)
+        if count >= limit:
+            descriptions = {
+                'send_email': _('You reached the limit of authentication mails sent for your account'),
+                'code_check': _('You reached the limit of code verifications for your account'),
+            }
+            description = descriptions.get(limit_type)
+            raise AccessDenied(description)
```

Comment:
It might be intentional (in order to avoid overloading the table?), but it looks like during rate limiting the hits are not being counted towards rate limiting. So a process which never stops will get locked out when it hits the rate limit, then unlocked after an hour has passed.

Reply:
This is a good point.

It's a choice between 
 - letting the attacker flooding a database table but unable to bruteforce until he stops flooding,
 - letting the attacker able to attempt 10 tries every one hour, but without flooding the table.

Since the code changes every hour, I would choose the second behavior, which will lead to less I/O without any significant security decrease. Olivier what do you think ?

Reply:
Yup, I'd favor the second option too, but we should add a comment so we remember that it was a conscious decision to reduce the chances of flood, trading that off with a limited "10 tries" window every hour.

However I wonder whether the limit per IP is sufficient. Botnets with thousands of IPs are increasingly common, and they could easily brute-force the code while staying under the per IP limit.

Some options that I see:
- 1) Simplest option: don't filter per IP while checking the limit, so a user whose login is under attack (even by a single IP) could be temporarily locked out. We could then allow the admin to reset the limit via an admin action that clears the logs for a given user.
- 2) Adaptive option: keep the per IP limit, but add an extra check about the number of different IPs that were used in the last hour. If the number of IPs is > 10, deny access too (with a slightly different message, so the admin can tell the difference) This way a user can't be locked out by a single IP, they need to be attacked by 10+ different IPs. Plus provide the same tool for an admin to reset the attempts. I think we can do this with a single SQL query still, as the IP filtering can be done in memory - in most cases there won't be many results.

Reply:
so this part of the rate-limiting logic will be taken into account in a follow-up PR, as discussed on the master PR #83750